File: solutions/rust/src/year2022/day04.rs
use std::ops::RangeInclusive;

use crate::{PartSolver, ProblemSolver};

type Pair = (RangeInclusive<u32>, RangeInclusive<u32>);

fn parse_range(input: &str) -> RangeInclusive<u32> {
    let [start, end] = input.split("-").next_chunk().unwrap();

    RangeInclusive::new(
        start.parse().expect(&format!("parse_range(): {input}")),
        end.parse().expect(&format!("parse_range(): {input}")),
    )
}

fn parse(input: &str) -> Vec<Pair> {
    let assignments = input.lines().map(|line| {
        let [first_elf, second_elf] = line.split(",").next_chunk().unwrap();
        (parse_range(first_elf), parse_range(second_elf))
    });

    assignments.collect()
}

fn count_pairs(assignments: &Vec<Pair>, predicate: impl FnMut(&&Pair) -> bool) -> i32 {
    assignments.iter().filter(predicate).count() as i32
}

// Part 1
trait FullyContains {
    fn fully_contains(&self, other: &Self) -> bool;
}

impl<T: Ord> FullyContains for RangeInclusive<T> {
    fn fully_contains(&self, other: &Self) -> bool {
        other.start() >= self.start() && other.end() <= self.end()
    }
}

fn part1(assignments: &Vec<Pair>) -> i32 {
    count_pairs(assignments, |(first_elf, second_elf)| {
        first_elf.fully_contains(second_elf) || second_elf.fully_contains(first_elf)
    })
}

// Part 2
trait Overlaps {
    fn overlaps(&self, other: &Self) -> bool;
}

impl<T: Ord> Overlaps for RangeInclusive<T> {
    fn overlaps(&self, other: &Self) -> bool {
        other.contains(&self.start()) || other.contains(&self.end())
    }
}

fn part2(assignments: &Vec<Pair>) -> i32 {
    count_pairs(assignments, |(first_elf, second_elf)| {
        first_elf.overlaps(second_elf) || second_elf.overlaps(first_elf)
    })
}

pub(crate) fn solution() -> ProblemSolver {
    ProblemSolver::new(PartSolver(parse, part1), PartSolver(parse, part2))
}

File: solutions/zig/src/year2022/day04.zig
const std = @import("std");

const RangeInclusive = struct {
    start: isize,
    end: isize,

    const Self = @This();

    pub inline fn contains(self: *const Self, item: isize) bool {
        return self.start <= item and item <= self.end;
    }

    pub inline fn fullyContains(self: *const Self, other: *const Self) bool {
        return other.start >= self.start and other.end <= self.end;
    }

    pub inline fn overlaps(self: *const Self, other: *const Self) bool {
        return other.contains(self.start) or other.contains(self.end);
    }
};

const Pair = struct { RangeInclusive, RangeInclusive };

fn parseRange(input: []const u8) !RangeInclusive {
    var members = std.mem.split(u8, input, "-");
    const start = try std.fmt.parseInt(isize, members.next().?, 10);
    const end = try std.fmt.parseInt(isize, members.next().?, 10);

    return .{
        .start = start,
        .end = end,
    };
}

fn parse(input: []const u8, allocator: std.mem.Allocator) !std.ArrayList(Pair) {
    var pairs = std.ArrayList(Pair).init(allocator);
    var lines = std.mem.split(u8, std.mem.trim(u8, input, " \n"), "\n");

    while (lines.next()) |line| {
        var pair = std.mem.split(u8, line, ",");
        try pairs.append(.{
            try parseRange(pair.next().?),
            try parseRange(pair.next().?),
        });
    }

    return pairs;
}

fn part1(assignments: *const std.ArrayList(Pair)) usize {
    var count: usize = 0;
    for (assignments.items) |pair| {
        if (pair[0].fullyContains(&pair[1]) or pair[1].fullyContains(&pair[0])) {
            count += 1;
        }
    }

    return count;
}

fn part2(assignments: *const std.ArrayList(Pair)) usize {
    var count: usize = 0;
    for (assignments.items) |pair| {
        if (pair[0].overlaps(&pair[1]) or pair[1].overlaps(&pair[0])) {
            count += 1;
        }
    }

    return count;
}

pub fn solution(input: []const u8, allocator: std.mem.Allocator) !struct { []u8, []u8 } {
    const assignments = try parse(input, allocator);

    return .{
        try std.fmt.allocPrint(allocator, "{}", .{part1(&assignments)}),
        try std.fmt.allocPrint(allocator, "{}", .{part2(&assignments)}),
    };
}
